docs/reference/PackageReference-with-packages-lock-file.md

```diff
@@ -0,0 +1,66 @@
+author: jainaashish
+ms.author: jainaashish
+ms.date: 10/18/2018
+ms.topic: conceptual
+---
+
+# Package references (PackageReference) with packages lock file (packages.lock.json)
+
+Package references, using the `PackageReference` node, manage NuGet dependencies directly within project files (as opposed to a separate `packages.config` file) [PackageReference in Project file](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files). 
+
+With PackageReference, Developers do not have confidence that NuGet will restore to the same full closure of package dependencies when they build it on Dev machine vs. CI/CD machines. They also doesn't know anything about their transitive dependencies coming through direct dependencies and when they get changed? (See for more details [Enable repeatable package restore using lock file](https://github.com/NuGet/Home/wiki/Enable-repeatable-package-restore-using-lock-file).)
+
+So with this packages lock file (`packages.lock.json`) feature, developers will be able to persist their restore resolution graph which should be added into their source control. Later this file will become the source of truth on CI/CD or any new machine to resolve the restore graph.
+
+## Opt-In to packages lock file
+
+Set a MSBuild property `RestorePackagesWithLockFile` in your project file using the following syntax:
+
+```xml
+<PropertyGroup>
+    <!--- ... -->
+    <RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>
+    <!--- ... -->
+</PropertyGroup>    
+```
+
+After that either run a restore or install/update a NuGet dependency, which will generate a `packages.lock.json` file at the project root level.
+
+You can also create blank `packages.lock.json` or `packages.<project_name>.lock.json` (replace spaces with `_` in project name) file at project root directory and run restore. This will also allow you to opt-in to this feature.
+
+## Overwriting default packages lock file path
+
+To overwrite default packages lock file name and path, Set MSBuild property `NuGetLockFilePath` in your project file using the following syntax:
+
+```xml
+<PropertyGroup>
+    <!--- ... -->
+    <NuGetLockFilePath>custom/<custom>.lock.json</NuGetLockFilePath>
+    <!--- ... -->
+</PropertyGroup>    
+```
+
+## Controlling restore locked mode
+
+By default restore will always update this packages lock file (if needed) but developer can control this by setting another MSBuild property `RestoreLockedMode` in the project file like below:
+
+```xml
+<PropertyGroup>
+    <!--- ... -->
+    <RestoreLockedMode>true</RestoreLockedMode>
+    <!--- ... -->
+</PropertyGroup>    
+```
+
+This property must be set to true as part of build scripts for CI/CD machines where restore should never update an existing lock file instead should fail so that developer can go back to their dev box, run restore which should update the lock file, test, and then check-in to the source control.
```

Comment:
Firstly, this sentence is too long and difficult to read. My suggestion:
This property must be set to true as part of build scripts for CI/CD machines where restore should fail when the restored packages do not match the lock file. Developers will need to check in an updated lock file to unblock the CI/CD build.

Secondly, I don't feel like editing an XML file (csproj) is a common task in a CI/CD build script. We can chat about it offline, but I feel like we should give an example on how to do it, even if we give an example that only works for one build system.

Reply:
I like the first suggestion and incorporate that.

About second part, yes nobody will go an update the csproj for CI/CD instead update their build script or build configuration. Idea here to tell which MSBuild property to set and its up to the developer to update their build accordingly, which is why the example about msbuild property. I dont know how to give an example for build script or configuration other than this. But I'm open for better suggestion.